### ISSUE TITLE
[BUG] Action forwarding is not working if Exception is handled in the Controller's initialize() phase

### ISSUE THREAD

**Issue by temuri416** [phalcon/cphalcon#12931]
Hi,

I use this approach to handle exceptions thrown in Controller Actions. Nothing fancy here:

```php
$eventsManager->attach('dispatch:beforeException', function($event, $dispatcher, $exception) {
    // Forward to Error controller:
    $dispatcher->forward([
        'controller' => 'error',
        'action' => 'oops',
        'params' => null,
    ]);
}
```

Everything works fine if an Exception is thrown inside Controller's Action, i.e. past this line:

https://github.com/phalcon/cphalcon/blob/master/phalcon/dispatcher.zep#L654

The problem is that if an Exception is thrown inside Controller's `initialize()`, then **forwarding code above does not work** (it does get executed).

Looking at the code I see something that looks like an oversight to me. This is the code that handles Exception in the named action ([link](https://github.com/phalcon/cphalcon/blob/master/phalcon/dispatcher.zep#L653)):

```php
try {
    // We update the latest value produced by the latest handler
    let this->_returnedValue = this->callActionMethod(handler, actionMethod, params);
} catch \Exception, e {
    if this->{"_handleException"}(e) === false {
        if this->_finished === false {
            continue;
        }
    } else {
        throw e;
    }
}
```

As you can see, an Exception is caught and `_handleException` is fired inside the dispatch loop, thus giving Dispatcher's `forward()` a chance  to affect execution in subsequent iteration.

And this is how Controller's `initialize` is called ([link](https://github.com/phalcon/cphalcon/blob/master/phalcon/dispatcher.zep#L600)):

```php
if wasFresh === true {
    if method_exists(handler, "initialize") {
        handler->initialize();
    }
    ...
}

```

So, if an Exception is thrown in `initialize`, the error bubbles up and out of the dispatch loop.

So, shouldn't there be a try\catch around the initialize() call as well?

Thanks!

### Details

* Phalcon version: 3.2.x
* PHP Version: (`php -v`) 7.1
* Operating System: Unix
* Installation type: Compiling from source
* Zephir version (if any): Latest

**Comment by virgofx** [comment]
I believe this is fixed with #12209  ... just need to rebase it and resubmit it. It fixes a bunch of looping and performance issues as well.

**Comment by temuri416** [comment]
will you be doing that any time soon?

**Comment by virgofx** [comment]
Yeah, it's on my todo list

**Comment by kowach** [comment]
I bumped on this bug too.

**Comment by NextSeason** [comment]
I hope you can fix this bug quickly, I can't continue doing my work because of this bug.

**Comment by temuri416** [comment]
@virgofx @sergeyklay 

Any update on this ticket? Is there some other dependency?

**Comment by virgofx** [comment]
@temuri416 Finally got around to this. Can you check https://github.com/virgofx/cphalcon/blob/80123d3628794ee937077c7a167c3c48730efcad/tests/unit/Mvc/Dispatcher/DispatcherInitalizeMethodTest.php#L180-L220   It should handle this specific use case.

In general, you can't use `forward()` directly in the initialize() method as this is essentially a constructor; however, you can forward as result of an exception caught via the `dispatch:beforeException` which is what this addresses.

**Comment by sergeyklay** [comment]
Fixed in the `3.2.x` branch. Feel free to open a new issue if the problem appears again. Thank you for contributing.
